### ISSUE TITLE
hsainfo utility

### ISSUE THREAD

**Issue by devurandom** [RadeonOpenCompute/ROCR-Runtime#15]
It would be nice to have a `hsainfo` utility that would show information about the HSA setup on the current system: Whether HSA is functional on my platform, what features it provides, and if it is not functional, why not.

In sample/vector_copy I get "Getting a gpu agent failed.". By hacking in a few printfs I already figured out that there seem to be no HSA devices available, but I am so far unable to figure out why that is - my kernel seems to detect everything just fine:

```
# uname -p
AMD A10-7800 Radeon R7, 12 Compute Cores 4C+8G
# dmesg | tail
[   75.584986] Found CRAT image with size=1440
[   75.584990] Parsing CRAT table with 1 nodes
[   75.584992] Found CU entry in CRAT table with proximity_domain=0 caps=0
[   75.584993] CU CPU: cores=4 id_base=16
[   75.584995] Found CU entry in CRAT table with proximity_domain=0 caps=0
[   75.584996] CU GPU: simds=32 id_base=-2147483648
[   75.584997] Found memory entry in CRAT table with proximity_domain=0
[   75.584999] Found memory entry in CRAT table with proximity_domain=0
[   75.585000] Found memory entry in CRAT table with proximity_domain=0
[   75.585001] Found memory entry in CRAT table with proximity_domain=0
[   75.585002] Found cache entry in CRAT table with processor_id=16
[   75.585003] Found cache entry in CRAT table with processor_id=16
[   75.585004] Found cache entry in CRAT table with processor_id=16
[   75.585005] Found cache entry in CRAT table with processor_id=17
[   75.585006] Found cache entry in CRAT table with processor_id=18
[   75.585006] Found cache entry in CRAT table with processor_id=18
[   75.585007] Found cache entry in CRAT table with processor_id=18
[   75.585008] Found cache entry in CRAT table with processor_id=19
[   75.585009] Found TLB entry in CRAT table (not processing)
[   75.585010] Found TLB entry in CRAT table (not processing)
[   75.585010] Found TLB entry in CRAT table (not processing)
[   75.585011] Found TLB entry in CRAT table (not processing)
[   75.585012] Found TLB entry in CRAT table (not processing)
[   75.585013] Found TLB entry in CRAT table (not processing)
[   75.585013] Found TLB entry in CRAT table (not processing)
[   75.585014] Found TLB entry in CRAT table (not processing)
[   75.585015] Found TLB entry in CRAT table (not processing)
[   75.585016] Found TLB entry in CRAT table (not processing)
[   75.585017] Creating topology SYSFS entries
[   75.585046] Finished initializing topology ret=0
[   75.587499] kfd kfd: Initialized module
# cat /sys/class/kfd/kfd/topology/nodes/0/properties
cpu_cores_count 4
simd_count 32
mem_banks_count 2
caches_count 3
io_links_count 0
cpu_core_id_base 16
simd_id_base 2147483648
max_waves_per_simd 40
lds_size_in_kb 64
gds_size_in_kb 0
wave_front_size 64
array_count 2
simd_arrays_per_engine 0
cu_per_simd_array 16
simd_per_cu 4
max_slots_scratch_cu 128
vendor_id 0
device_id 0
location_id 0
max_engine_clk_ccompute 3500
```

**Comment by gstoner** [comment]
Great suggestion,  .  Can you tell us make and brand of the PC your  working with.

When you working with Carizzo system we have to have two things working in the system BIOS

IOMMUv2 needs to be turned on
And the BIOS  need the CRAT table correctly set -  it looks like the Vendor did not set the CRAT table properly.

Greg



On Aug 20, 2016, at 8:42 AM, Dennis Schridde <notifications@github.com<mailto:notifications@github.com>> wrote:


It would be nice to have a hsainfo utility that would show information about the HSA setup on the current system: Whether HSA is functional on my platform, what features it provides, and if it is not functional, why not.

In sample/vector_copy I get "Getting a gpu agent failed.". By hacking in a few printfs I already figured out that there seem to be no HSA devices available, but I am so far unable to figure out why that is - my kernel seems to detect everything just fine:

# uname -p
AMD A10-7800 Radeon R7, 12 Compute Cores 4C+8G
# dmesg | tail
[   75.584986] Found CRAT image with size=1440
[   75.584990] Parsing CRAT table with 1 nodes
[   75.584992] Found CU entry in CRAT table with proximity_domain=0 caps=0
[   75.584993] CU CPU: cores=4 id_base=16
[   75.584995] Found CU entry in CRAT table with proximity_domain=0 caps=0
[   75.584996] CU GPU: simds=32 id_base=-2147483648
[   75.584997] Found memory entry in CRAT table with proximity_domain=0
[   75.584999] Found memory entry in CRAT table with proximity_domain=0
[   75.585000] Found memory entry in CRAT table with proximity_domain=0
[   75.585001] Found memory entry in CRAT table with proximity_domain=0
[   75.585002] Found cache entry in CRAT table with processor_id=16
[   75.585003] Found cache entry in CRAT table with processor_id=16
[   75.585004] Found cache entry in CRAT table with processor_id=16
[   75.585005] Found cache entry in CRAT table with processor_id=17
[   75.585006] Found cache entry in CRAT table with processor_id=18
[   75.585006] Found cache entry in CRAT table with processor_id=18
[   75.585007] Found cache entry in CRAT table with processor_id=18
[   75.585008] Found cache entry in CRAT table with processor_id=19
[   75.585009] Found TLB entry in CRAT table (not processing)
[   75.585010] Found TLB entry in CRAT table (not processing)
[   75.585010] Found TLB entry in CRAT table (not processing)
[   75.585011] Found TLB entry in CRAT table (not processing)
[   75.585012] Found TLB entry in CRAT table (not processing)
[   75.585013] Found TLB entry in CRAT table (not processing)
[   75.585013] Found TLB entry in CRAT table (not processing)
[   75.585014] Found TLB entry in CRAT table (not processing)
[   75.585015] Found TLB entry in CRAT table (not processing)
[   75.585016] Found TLB entry in CRAT table (not processing)
[   75.585017] Creating topology SYSFS entries
[   75.585046] Finished initializing topology ret=0
[   75.587499] kfd kfd: Initialized module
# cat /sys/class/kfd/kfd/topology/nodes/0/properties
cpu_cores_count 4
simd_count 32
mem_banks_count 2
caches_count 3
io_links_count 0
cpu_core_id_base 16
simd_id_base 2147483648
max_waves_per_simd 40
lds_size_in_kb 64
gds_size_in_kb 0
wave_front_size 64
array_count 2
simd_arrays_per_engine 0
cu_per_simd_array 16
simd_per_cu 4
max_slots_scratch_cu 128
vendor_id 0
device_id 0
location_id 0
max_engine_clk_ccompute 3500

**Comment by devurandom** [comment]
The processor is an AMD A10-7800 (see `uname -p` above) - a Kaveri with 2x2 CPU Cores and 8 GPU Compute Units. I bought it back then, because it was the first to support HSA (2.0).

**Comment by devurandom** [comment]
`dmesg` appears to show that IOMMUv2 is present and detected:
```
[    0.945322] AMD-Vi: IOMMU performance counters supported
[    0.945328] AMD-Vi: Applying ATS write check workaround for IOMMU at 0000:00:00.2
[    0.945867] iommu: Adding device 0000:00:01.0 to group 0
[    0.945878] iommu: Using direct mapping for device 0000:00:01.0
[    0.945893] iommu: Adding device 0000:00:01.1 to group 0
[    0.946192] iommu: Adding device 0000:00:02.0 to group 1
[    0.946552] iommu: Adding device 0000:00:03.0 to group 2
[    0.946882] iommu: Adding device 0000:00:04.0 to group 3
[    0.947220] iommu: Adding device 0000:00:10.0 to group 4
[    0.947236] iommu: Adding device 0000:00:10.1 to group 4
[    0.947563] iommu: Adding device 0000:00:11.0 to group 5
[    0.947893] iommu: Adding device 0000:00:12.0 to group 6
[    0.947907] iommu: Adding device 0000:00:12.2 to group 6
[    0.948237] iommu: Adding device 0000:00:13.0 to group 7
[    0.948250] iommu: Adding device 0000:00:13.2 to group 7
[    0.948623] iommu: Adding device 0000:00:14.0 to group 8
[    0.948637] iommu: Adding device 0000:00:14.2 to group 8
[    0.948650] iommu: Adding device 0000:00:14.3 to group 8
[    0.948983] iommu: Adding device 0000:00:14.4 to group 9
[    0.949332] iommu: Adding device 0000:00:15.0 to group 10
[    0.949348] iommu: Adding device 0000:00:15.1 to group 10
[    0.949364] iommu: Adding device 0000:00:15.2 to group 10
[    0.949740] iommu: Adding device 0000:00:18.0 to group 11
[    0.949755] iommu: Adding device 0000:00:18.1 to group 11
[    0.949769] iommu: Adding device 0000:00:18.2 to group 11
[    0.949783] iommu: Adding device 0000:00:18.3 to group 11
[    0.949797] iommu: Adding device 0000:00:18.4 to group 11
[    0.949812] iommu: Adding device 0000:00:18.5 to group 11
[    0.949822] iommu: Adding device 0000:03:00.0 to group 10
[    0.949835] iommu: Adding device 0000:04:00.0 to group 10
[    0.950253] AMD-Vi: Found IOMMU at 0000:00:00.2 cap 0x40
[    0.950255] AMD-Vi:  Extended features:  PPR GT IA PC
[    0.950258] AMD-Vi: Interrupt remapping enabled
[    0.950484] AMD-Vi: Lazy IO/TLB flushing enabled
[    0.951741] perf: AMD NB counters detected
[    0.951823] LVT offset 0 assigned for vector 0x400
[    0.951834] perf: AMD IBS detected (0x000001ff)
[    0.951838] perf: amd_iommu: Detected. (0 banks, 0 counters/bank)
[    0.981105] AMD IOMMUv2 driver by Joerg Roedel <jroedel@suse.de>
```

My mainboard is an ASUS A88X-PRO, which I just updated to firmware version 2603.

**Comment by devurandom** [comment]
/dev/kfd was owned by 0:0 and had mode 0600 - thus it was not accessible by the current user. How would I check for this problem in source code?

After making it accessible, I get:
```
vector_copy: topology.c:671: topology_sysfs_get_node_props: Assertion `props->EngineId.ui32.Major' failed.
fish: “./vector_copy” terminated by signal SIGABRT (Abort)
```

The source code is not really verbose on what it tries to assert.

**Comment by gstoner** [comment]
One thing Kaveri is a Pre-HSA Compliant (not HSA 1.0 compliant)  APU, which we used as a development vehicle in preparation of the HSA 1.0 compliant APU's Carizzo.  We used it to get a bulk of the feature ready.   

We used the same motherboard for development before getting Carizzo development systems.    Now I know the motherboard and APU part.  Are you trying to run ROCm 1.2?

Also, I will send you a private message about and idea around equipment

**Comment by devurandom** [comment]
> One thing Kaveri is a Pre-HSA Compliant (not HSA 1.0 compliant) APU, which we used as a development vehicle in preparation of the HSA 1.0 compliant APU's Carizzo. We used it to get a bulk of the feature ready. 

[Wikipedia claims it supports "HSA 2.0"](https://en.wikipedia.org/wiki/List_of_AMD_accelerated_processing_unit_microprocessors#.22Kaveri.22_.282014.2C_28_nm.29)... Should I fixed the article?

> We used the same motherboard for development before getting Carizzo development systems. Now I know the motherboard and APU part. Are you trying to run ROCm 1.2?

Yes, I built libhsakmt.so and libhsa-runtime64.so from the ROC 1.2.0 sources. The sample code is from the latest git of ROCR-Runtime.

> Also, I will send you a private message about and idea around equipment 

Yes, please. You can find my email address on my GH user page.

**Comment by gstoner** [comment]
There is no such thing.. There is the HSA 1.0 and HSA 1.1 specification today.  Next spec is 1.2   I am Chairman of the Board and Managing Director of HSA Foundation

HSA 1.1 Specifications - web based documents
http://www.hsafoundation.com/html_spec11/HSA_Library.htm

HSA 1.0 Specifications web based documents
http://www.hsafoundation.com/html/HSA_Library.htm

Also email sent.


On Aug 20, 2016, at 1:25 PM, Dennis Schridde <notifications@github.com> wrote:


One thing Kaveri is a Pre-HSA Compliant (not HSA 1.0 compliant) APU, which we used as a development vehicle in preparation of the HSA 1.0 compliant APU's Carizzo. We used it to get a bulk of the feature ready.

Wikipedia claims it supports "HSA 2.0"<https://en.wikipedia.org/wiki/List_of_AMD_accelerated_processing_unit_microprocessors#.22Kaveri.22_.282014.2C_28_nm.29>... Should I fixed the article?

We used the same motherboard for development before getting Carizzo development systems. Now I know the motherboard and APU part. Are you trying to run ROCm 1.2?

Yes, I built libhsakmt.so and libhsa-runtime64.so from the ROC 1.2.0 sources. The sample code is from the latest git of ROCR-Runtime.

Also, I will send you a private message about and idea around equipment

Yes, please. You can find my email address on my GH user page.

**Comment by gstoner** [comment]
We built a tool like this rocminfo

**Comment by gogoprog** [comment]
> /dev/kfd was owned by 0:0 and had mode 0600 - thus it was not accessible by the current user. How would I check for this problem in source code?
> 
> After making it accessible, I get:
> 
> ```
> vector_copy: topology.c:671: topology_sysfs_get_node_props: Assertion `props->EngineId.ui32.Major' failed.
> fish: “./vector_copy” terminated by signal SIGABRT (Abort)
> ```
> 
> The source code is not really verbose on what it tries to assert.

I have the same issue while running `clinfo` or `rocminfo` and I don't know what to do next

**Comment by fxkamd** [comment]
/dev/kfd permissions are controlled by udevd. You can add a udev rule to set the permissions correctly. E.g. paste this into /etc/udev/rules.d/70-kfd.rules:

SUBSYSTEM=="kfd", KERNEL=="kfd", TAG+="uaccess", GROUP="video"

The assertion failure looks like your GPU is not supported by the Thunk. This assert should only happen when there is no entry for your GPU's device ID in gfxip_lookup_table. What GPU are you using, and what ROCm version? Can you provide the output of these commands:

lspci -d 1002: -nn
for x in /sys/class/kfd/kfd/topology/nodes/*/properties; do echo "=== $x ==="; cat $x; done

**Comment by gogoprog** [comment]
Hello,

Thanks for the answer.
I have a 6700 XT.

```
# lspci -d 1002: -nn                                                                                                                                                                                         !575
0a:00.0 PCI bridge [0604]: Advanced Micro Devices, Inc. [AMD/ATI] Navi 10 XL Upstream Port of PCI Express Switch [1002:1478] (rev c5)
0b:00.0 PCI bridge [0604]: Advanced Micro Devices, Inc. [AMD/ATI] Navi 10 XL Downstream Port of PCI Express Switch [1002:1479]
0c:00.0 VGA compatible controller [0300]: Advanced Micro Devices, Inc. [AMD/ATI] Navi 22 [1002:73df] (rev c5)
0c:00.1 Audio device [0403]: Advanced Micro Devices, Inc. [AMD/ATI] Device [1002:ab28]
```

```
#for x in /sys/class/kfd/kfd/topology/nodes/*/properties; do echo "=== $x ==="; cat $x; done
=== /sys/class/kfd/kfd/topology/nodes/0/properties ===
cpu_cores_count 32
simd_count 0
mem_banks_count 1
caches_count 0
io_links_count 0
cpu_core_id_base 0
simd_id_base 0
max_waves_per_simd 0
lds_size_in_kb 0
gds_size_in_kb 0
num_gws 0
wave_front_size 0
array_count 0
simd_arrays_per_engine 0
cu_per_simd_array 0
simd_per_cu 0
max_slots_scratch_cu 0
vendor_id 0
device_id 0
location_id 0
domain 0
drm_render_minor 0
hive_id 0
num_sdma_engines 0
num_sdma_xgmi_engines 0
num_sdma_queues_per_engine 0
num_cp_queues 0
max_engine_clk_ccompute 3400
=== /sys/class/kfd/kfd/topology/nodes/1/properties ===
cpu_cores_count 0
simd_count 160
mem_banks_count 1
caches_count 60
io_links_count 1
cpu_core_id_base 0
simd_id_base 2147487744
max_waves_per_simd 16
lds_size_in_kb 64
gds_size_in_kb 0
num_gws 0
wave_front_size 32
array_count 4
simd_arrays_per_engine 2
cu_per_simd_array 10
simd_per_cu 4
max_slots_scratch_cu 32
vendor_id 4098
device_id 29663
location_id 3072
domain 0
drm_render_minor 128
hive_id 0
num_sdma_engines 2
num_sdma_xgmi_engines 0
num_sdma_queues_per_engine 8
num_cp_queues 8
max_engine_clk_fcompute 2725
local_mem_size 0
fw_version 73
capability 8832
sdma_fw_version 63
unique_id 0
max_engine_clk_ccompute 3400

```

**Comment by fxkamd** [comment]
@gogoprog, the next ROCm release should enable this GPU.

**Comment by gogoprog** [comment]
@fxkamd Thank you

**Comment by gogoprog** [comment]
Same issue with ROCm 4.2.0

**Comment by gogoprog** [comment]
> Same issue with ROCm 4.2.0

Fixed after installing hsakmt-roct 4.2.0
